pkg/profiles/tap.go

```diff
@@ -0,0 +1,142 @@
+package profiles
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/linkerd/linkerd2/controller/api/util"
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
+	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// RenderTapOutputProfile performs a tap on the desired resource and generates
+// a service profile with routes pre-populated from the tap data
+// Only inbound tap traffic is considered.
+func RenderTapOutputProfile(client pb.ApiClient, tapResource, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int, w io.Writer) error {
+	requestParams := util.TapRequestParams{
+		Resource:  tapResource,
+		Namespace: namespace,
+	}
+	log.Debugf("Running `linkerd tap %s --namespace %s`", tapResource, namespace)
+
+	req, err := util.BuildTapByResourceRequest(requestParams)
+	if err != nil {
+		return err
+	}
+
+	profile, err := tapToServiceProfile(client, req, namespace, name, controlPlaneNamespace, tapDuration, routeLimit)
+	if err != nil {
+		return err
+	}
+
+	output, err := yaml.Marshal(profile)
+	if err != nil {
+		return fmt.Errorf("Error writing Service Profile: %s", err)
+	}
+	w.Write(output)
+	return nil
+}
+
+func tapToServiceProfile(client pb.ApiClient, tapReq *pb.TapByResourceRequest, namespace, name, controlPlaneNamespace string, tapDuration time.Duration, routeLimit int) (sp.ServiceProfile, error) {
+	profile := sp.ServiceProfile{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      fmt.Sprintf("%s.%s.svc.cluster.local", name, namespace),
+			Namespace: controlPlaneNamespace,
+		},
+		TypeMeta: meta_v1.TypeMeta{
+			APIVersion: "linkerd.io/v1alpha1",
+			Kind:       "ServiceProfile",
+		},
+	}
+
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(tapDuration))
+	defer cancel()
+
+	tapClient, err := client.TapByResource(context.Background(), tapReq)
+	if err != nil {
+		return profile, err
+	}
+
+	routes := routeSpecFromTap(ctx, tapClient, routeLimit)
+
+	profile.Spec.Routes = routes
+
+	return profile, nil
+}
+
+func routeSpecFromTap(ctx context.Context, tapClient pb.Api_TapByResourceClient, routeLimit int) []*sp.RouteSpec {
+	routes := make([]*sp.RouteSpec, 0)
+	routesMap := make(map[string]*sp.RouteSpec)
+
+recvLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			break recvLoop
+		default:
+			log.Debug("Waiting for data...")
```

Comment:
What do you think about removing this log line and adding a debug log line that displays the routeSpec this function was able to create after receiving the tap event. This might provide useful information in the case where we need to observe what routeSpecs are being created based on the data we are getting from tap.

Reply:
I think having the routespec data could be useful! I think this is here so we know we've started a tap, but we are waiting to receive (could be useful if we're waiting a while without any results). added another log line for logging the routespec.